cpu/atmega_common/periph/rtt.c

```diff
@@ -39,28 +54,99 @@
 #define ENABLE_DEBUG    (0)
 #include "debug.h"
 
+#ifdef RTT_BACKEND_SC
+/**
+ * Atomic read of 32-bit registers according
+ * to section 10.3 of the Datasheet
+ * @{
+ */
+
+__attribute__((always_inline))
+  static inline uint32_t rg_read32(volatile uint8_t *hh,
+                                   volatile uint8_t *hl,
+                                   volatile uint8_t *lh,
+                                   volatile uint8_t *ll)
+{
+    le_uint32_t rg;
+
+    rg.u8[0] = *ll;
+    rg.u8[1] = *lh;
+    rg.u8[2] = *hl;
+    rg.u8[3] = *hh;
+
+    return rg.u32;
+}
+
+__attribute__((always_inline))
+  static inline void rg_write32(uint32_t val, volatile uint8_t *hh,
+                                              volatile uint8_t *hl,
+                                              volatile uint8_t *lh,
+                                              volatile uint8_t *ll)
+{
+    le_uint32_t rg;
+
+    rg.u32 = val;
+
+    *hh = rg.u8[3];
+    *hl = rg.u8[2];
+    *lh = rg.u8[1];
+    *ll = rg.u8[0];
+}
```

Comment:
The target should be the first parameter. (E.g. see `memcpy()`, `strcpy()`, `read()`, `recv()`, `recvfrom()`, `recvmsg()`, etc. Only file descriptors or library handles come prior to the target.)

Again, only the address of the least significant byte needs to be known, as the rest is consecutive:

```C
void reg32_write(volatile uint8_t *reg_ll, uint32_t _val)
{
    le_uint32_t val = { .u32 = _val };
    reg_ll[3] = val.u8[3];
    reg_ll[2] = val.u8[2];
    reg_ll[1] = val.u8[1];
    reg_ll[0] = val.u8[0];
}
```

Reply:
Ok.

Reply:
This is also only used where interrupts are already disabled - except for `_timer_init()`, but we could just disable the interrupts there and keep the helper functions simple.